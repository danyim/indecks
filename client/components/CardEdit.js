import React from 'react';
import { Link, browserHistory } from 'react-router';
import CSSTransitionGroup from 'react-addons-css-transition-group';
import styles from '../styles/components/CardEdit';

const CardEdit = React.createClass({
  title: null,
  answer: null,
  cardIndex: null,
  deckId: null,

  // Not sure if this is the right path...
  // The text fields are not updating
  handleChangeTitle(e) {
    // this.setState({
    //   deck
    // })
  },
  handleChangeAnswer(e) {

  },
  handleSubmit(e) {
    e.preventDefault();
    this.props.CardEdit(
      this.refs.title.value,
      this.refs.answer.value,
      this.cardIndex,
      this.deckId
    );
    browserHistory.push(`/view/${this.deckId}/${this.cardIndex}`)
  },
  render() {
    const { title = '', answer = '' } = this.props.card;
    this.deckId = this.props.deckId || this.params.deckId;
    this.cardIndex = this.props.cardIndex;

    return (
      <figure className={`grid-figure ${styles['grid-figure']}`}>
        <form ref="commentForm" className="edit-form" onSubmit={this.handleSubmit}>
          <input type="text" className="large-input" name="title" ref="title" placeholder="Title" defaultValue={title} />
          <textarea type="text" className="mono" name="answer" ref="answer" placeholder="Answer (Markdown)" defaultValue={answer} rows="6" />
          <button type="submit" className="button">Save</button>
        </form>
      </figure>
    );
  }
})

export default CardEdit;
