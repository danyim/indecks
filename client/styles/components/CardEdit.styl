.grid-figure
  padding 2rem
  button
    margin-top 1rem
