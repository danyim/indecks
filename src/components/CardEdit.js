import React from 'react';
import { Link, browserHistory } from 'react-router';
import CSSTransitionGroup from 'react-addons-css-transition-group';
import styles from '../styles/components/CardEdit';

class CardEdit extends React.Component {
  constructor(props) {
    super(props);
    this.title = null;
    this.answer = null;
    this.cardIndex = null;
    this.deckId = null;

    this.checkIfDirty = this.checkIfDirty.bind(this);
    this.handleDelete = this.handleDelete.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  // Not sure if this is the right path...
  // The text fields are not updating
  handleChangeTitle(e) {
    // this.setState({
    //   deck
    // })
  }

  handleChangeAnswer(e) {

  }

  checkIfDirty() {
    const card = this.props.card;
    return (this.refs.title.value !== card.title || this.refs.answer.value !== card.answer);
  }

  handleCancel() {
    if(this.checkIfDirty()) {
      if(confirm('Are you sure? You have unsaved changes.')) {
        browserHistory.push(`/view/${this.deckId}`);
      }
    }
    else {
      browserHistory.push(`/view/${this.deckId}/${this.cardIndex}`);
    }
  }

  handleDelete(cardIndex, deckId) {
    if(confirm('Are you sure?')) {
      this.props.removeCard(
        cardIndex,
        deckId
      );
      browserHistory.push(`/view/${this.deckId}`);
    }
  }

  handleSubmit(e) {
    e.preventDefault();
    this.props.editCard(
      this.refs.title.value,
      this.refs.answer.value,
      this.cardIndex,
      this.deckId
    );
    browserHistory.push(`/view/${this.deckId}/${this.cardIndex}`)
  }

  render() {
    const { title = '', answer = '' } = this.props.card;
    this.deckId = this.props.deckId || this.params.deckId;
    this.cardIndex = this.props.cardIndex;

    return (
      <figure className={`grid-figure ${styles['grid-figure']}`}>
        <form ref="commentForm" className="edit-form" onSubmit={this.handleSubmit}>
          <textarea className="large-input" name="title" ref="title" placeholder="Title" defaultValue={title} rows="2" />
          <textarea className="mono" name="answer" ref="answer" placeholder="Answer (Markdown)" defaultValue={answer} rows="6" />
          <p><a>Preview Changes</a></p>
          <button type="submit" className="button">Save Card</button>
          <button type="button" className="button" onClick={() => this.handleCancel()}>Cancel</button>
          <button type="button" className="button btn-delete" onClick={() => this.handleDelete(this.cardIndex, this.deckId)}>Remove from Deck</button>
        </form>
      </figure>
    );
  }
}

export default CardEdit;
