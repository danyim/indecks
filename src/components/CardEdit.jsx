import React from 'react'
import { browserHistory } from 'react-router'
import KeyBinding from 'react-keybinding-component'
import CardEditForm from './CardEditForm'
import styles from '../styles/components/CardEdit.styl'

const propTypes = {
  card: React.PropTypes.object.isRequired,
  cardIndex: React.PropTypes.number.isRequired,
  deckId: React.PropTypes.string.isRequired,
  editCard: React.PropTypes.func.isRequired,
  removeCard: React.PropTypes.func.isRequired,
  form: React.PropTypes.object.isRequired
}

const defaultProps = {}

class CardEdit extends React.Component {
  constructor (props) {
    super(props)

    this.checkIfDirty = this.checkIfDirty.bind(this)
    this.handleCancel = this.handleCancel.bind(this)
    this.handleDelete = this.handleDelete.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  checkIfDirty () {
    const { title, answer } = this.props.card
    return (
      this.props.form.cardEdit.values.cardTitle !== title ||
      this.props.form.cardEdit.values.cardAnswer !== answer)
  }

  handleKeyDown (e) {
    if (e.keyCode === 27) { // escape
      this.handleCancel()
    }

    e.stopPropagation()
  }

  handleCancel () {
    if (this.checkIfDirty()) {
      if (confirm('Are you sure you want to cancel? You have unsaved changes that will be lost.')) {
        browserHistory.push(`/view/${this.props.deckId}`)
      }
    } else {
      browserHistory.push(`/view/${this.props.deckId}/${this.props.cardIndex}`)
    }
  }

  handleDelete (cardIndex, deckId) {
    if (confirm('Are you sure?')) {
      this.props.removeCard(
        cardIndex,
        deckId
      )
      browserHistory.push(`/view/${this.props.deckId}`)
    }
  }

  handleSubmit (e) {
    e.preventDefault()
    this.props.editCard(
      this.props.form.cardEdit.values.cardTitle,
      this.props.form.cardEdit.values.cardAnswer,
      this.props.cardIndex,
      this.props.deckId
    )
    browserHistory.push(`/view/${this.props.deckId}/${this.props.cardIndex}`)
  }

  render () {
    const { card, deckId, cardIndex } = this.props
    const { title = '', answer = '' } = card

    const initialValues = {
      cardTitle: title,
      cardAnswer: answer
    }

    return (
      <section className='single'>
        <figure className={`grid-figure ${styles['grid-figure']}`}>
          <KeyBinding onKey={e => this.handleKeyDown(e)} />
          <CardEditForm
            deckId={deckId}
            cardIndex={cardIndex}
            title={title}
            answer={answer}
            formValues={this.props.form.cardEdit}
            initialValues={initialValues}
            handleSubmit={this.handleSubmit}
            handleDelete={this.handleDelete}
            handleCancel={this.handleCancel}
          />
        </figure>
      </section>
    )
  }
}

CardEdit.propTypes = propTypes
CardEdit.defaultProps = defaultProps

export default CardEdit
