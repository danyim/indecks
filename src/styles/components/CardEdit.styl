
@import '../_variables'

// figure is added here to give precedence over the base grid-figure
figure.grid-figure
  display flex
  min-height 275px
  flex-flow column nowrap
  justify-content: center

.card-title p
  font-size 2.5rem
  padding 0.5rem 0rem

.card-answer p, li
  font-size 1.5rem
  line-height 1.5

.preview p
  font-weight bold

.preview-pane
  padding 1.0rem 2.5rem
  background-color lighten(lightgrey, 90%)
  border 1px solid lightgrey


.control-buttons
  flex: 0 0 auto
  flex-flow: row nowrap
  margin-top: 1.5rem
  button
    margin 0 0 0 1rem
  button:first-child
    margin 0

@media (max-width: cardWidth)
  .control-buttons
    flex: 0 0 50%
    flex-flow: column nowrap
    align-content space-around
    button
      margin 0.6rem 0 0 0
    button:first-child
      margin 0
