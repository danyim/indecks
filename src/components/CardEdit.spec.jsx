import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import CardEdit from './CardEdit';

const defaultProps = {
  card: {},
  cardIndex: 1,
  deckId: 'ABCDEFG',
  editCard: () => {},
  removeCard: () => {},
  form: {}
};

function setup(props = defaultProps) {
  const wrapper = shallow(<CardEdit {...props} />)

  return {
    props,
    wrapper
  }
}

describe('CardEdit', () => {
  // it('should render self and subcomponents', () => {
  //   const { wrapper } = setup();
  //   expect(wrapper.find('section.single').exists()).toBe(true);

  //   const tree = renderer.create(
  //     <CardEdit {...defaultProps} />
  //   ).toJSON();
  //   expect(tree).toMatchSnapshot();
  // });
});
