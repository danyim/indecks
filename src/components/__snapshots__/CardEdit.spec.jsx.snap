// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardAdd should render self and subcomponents 1`] = `
<figure
  className="grid-figure grid-figure"
>
  <form
    className="edit-form"
    onSubmit={[Function]}
  >
    <input
      className="large-input"
      name="title"
      onChange={[Function]}
      placeholder="Title"
      type="text"
    />
    <textarea
      className="mono"
      name="answer"
      onChange={[Function]}
      placeholder="Answer (Markdown)"
      rows="4"
      type="text"
    />
    <div
      className="preview"
    >
      <button
        className="btn pointer m-t"
        onClick={[Function]}
        type="button"
      >
        Preview [
        +
        ]
      </button>
      <div
        className="preview-pane"
        style={
          Object {
            "display": "none",
            "height": 0,
          }
        }
      >
        <div
          className="card-title"
          dangerouslySetInnerHTML={
            Object {
              "__html": "",
            }
          }
          onClick={null}
        />
        <hr />
        <div
          className="card-answer"
          dangerouslySetInnerHTML={
            Object {
              "__html": "",
            }
          }
          onClick={null}
        />
      </div>
    </div>
    <div
      className="control-buttons"
    >
      <button
        className="button"
        type="submit"
      >
        Save
      </button>
      <button
        className="button"
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</figure>
`;
